### PR TITLE
Implemented posting Twitch Live notifications to Discord Channels 

### LINKED ISSUE #14: Notifications for streamers on the FragSoc channel
Bot should send a notification message whenever each of our community streamers goes live on the FragSoc twitch channel

### PR DESCRIPTION
**Pull request by Fluxticks**
Implemented use of Twitch API and webhooks to receive notifications of when streamers go live. 

This resolves #14

### CODE REVIEW

**Comment by Laura7089** [comment]
You'll need to merge from master

**Comment by Fluxticks** [comment]
Am I meant to be merging straight into master?

**Comment by Laura7089** [comment]
Sorry, `develop` - you'll need to merge develop into this to get ryan's database changes.

**Comment by Fluxticks** [comment]
Oh yeah I did that already with commit #[0269886](https://github.com/FragSoc/esports-bot/pull/140/commits/02698866d6a693b6905f38f9da8ec25234c4616a)

File: src/esportsbot/cogs/TwitchCog.py [hunk 1]
```diff
@@ -0,0 +1,962 @@
...
+        """
+
+        if channel_id := (await self.get_channel_id(ctx, channel)):
+
+            db_entry = self._db.get(Twitch_info, guild_id=ctx.guild.id, twitch_channel_id=channel_id)
+            if db_entry is None:
+                await ctx.send(self.user_strings["channel_not_added_error"].format(channel=channel))
+                return
+
+            if message is not None and message.strip() == "" or message == "":
```
Comment (Laura7089) [inline src/esportsbot/cogs/TwitchCog.py:806]: The last part of this bool isn't necessary

File: src/esportsbot/cogs/TwitchCog.py [hunk 2]
```diff
@@ -0,0 +1,962 @@
...
+            name=channel_info.get("broadcaster_login"),
+            url=f"https://www.twitch.tv/{channel_info.get('broadcaster_login')}",
+            icon_url=channel_info.get("thumbnail_url")
+        )
+        embed.set_thumbnail(url=channel_info.get("thumbnail_url"))
+        embed.add_field(name="Current Game:", value=f"{channel_info.get('game_name')}")
+
+        await ctx.send(embed=embed)
+
+    # TODO: Probably best to move this to lib or some other as it is shared by TwitterCog
```
Comment (Laura7089) [inline src/esportsbot/cogs/TwitchCog.py:880]: Still relevant?

File: src/esportsbot/cogs/TwitchCog.py [hunk 3]
```diff
@@ -0,0 +1,962 @@
...
+        channel = self._bot.get_channel(cleaned_id)
+        if channel is None:
+            try:
+                channel = await self._bot.fetch_channel(cleaned_id)
+            except Forbidden as e:
+                self.logger.error("Unable to access channel with id %s due to permission errors: %s", cleaned_id, e.text)
+                return None
+        return channel
+
+    # TODO: Probably best to move this to lib or some other as it is shared by TwitterCog
```
Comment (Laura7089) [inline src/esportsbot/cogs/TwitchCog.py:903]: Again, TODO still relevant?

**Review by Laura7089 — CHANGES_REQUESTED** [review]
Awesome stuff. Will the web server need a port forwarding? Is the README up to date?

File: src/esportsbot/cogs/TwitchCog.py [hunk 4]
```diff
@@ -0,0 +1,962 @@
...
+                    data = await response.json()
+                    self.bearer = {
+                        "granted_on": current_time.strftime(DATETIME_FMT),
+                        "expires_in": data.get("expires_in"),
+                        "access_token": data.get("access_token")
+                    }
+
+        # TODO: Remove temp code
+        if self.bearer is not None:
+            with open("bearer", "w") as f:
```
[the remark this answers is not in the document]
Comment (Laura7089) [reply src/esportsbot/cogs/TwitchCog.py:117]: Perhaps https://docs.python.org/3/library/tempfile.html? Depends how long you want the file to live really

**Comment by Laura7089** [comment]
Closes #14

File: src/esportsbot/cogs/TwitchCog.py [hunk 5]
```diff
@@ -0,0 +1,962 @@
...
+        Is executed whenever the client loses a connection to Discord. Could be when no internet or when logged out.
+        """
+
+        if self._bot.is_closed:
+            self._http_server.stop()
+
+    @commands.Cog.listener()
+    async def on_webhooks_update(self, channel):
+        pass
+        # TODO: Capture this event to determine when a webhook gets deleted not using the command.
```
[the remark this answers is not in the document]
Comment (Fluxticks) [reply src/esportsbot/cogs/TwitchCog.py:495]: Yeah I haven't figured out how to properly make use of this event but it is something that would be good to implement.

**Comment by Fluxticks** [comment]
I implemented most, if not all, of the changes you suggested, thanks :)

File: src/esportsbot/cogs/TwitchCog.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Laura7089) [reply src/esportsbot/cogs/TwitchCog.py:495]: Cool, if there's nothing to be done immediately we're probably fine committing this as-is

**Review by Laura7089 — APPROVED** [review]

**Comment by Laura7089** [comment]
> Awesome stuff. Will the web server need a port forwarding? Is the README up to date?

Bump on these

**Comment by Fluxticks** [comment]
README is up to date yes, and yes it will need portforwarding

**Comment by Laura7089** [comment]
> it will need portforwarding

You'll want to open the port in the docker-compose file. I can do that if you want (what port)?

**Comment by Fluxticks** [comment]
Yeah it currently binds itself to port 443, but if you could do the docker port thing for me that'd be great :)

**Comment by Laura7089** [comment]
Sorry, I never saw your reply :(
Does this need TLS?

**Comment by Fluxticks** [comment]
Er, my networking knowledge is very limited so not sure. I used ngrok for all of the network stuff while developing so all I know is the port number

**Comment by Laura7089** [comment]
If it worked during your testing with the official twitch api, it's probably fine, I'll take a look at the official doc to be doubly sure

**Comment by Laura7089** [comment]
Scratch that, it does need https as mentioned here https://dev.twitch.tv/docs/eventsub

**Pull request merged**